crates/red_knot_python_semantic/src/types/context.rs

```diff
@@ -112,6 +126,42 @@ impl<'db> InferContext<'db> {
         });
     }
 
+    pub(super) fn set_in_no_type_check(&mut self, no_type_check: InNoTypeCheck) {
+        self.no_type_check = no_type_check;
+    }
```

Comment:
It seems like we currently only ever set this field to the `Yes` variant, so I'd be tempted to do something like this, which feels more readable to me:

```diff
diff --git a/crates/red_knot_python_semantic/src/types/context.rs b/crates/red_knot_python_semantic/src/types/context.rs
index 9543bde55..ace859425 100644
--- a/crates/red_knot_python_semantic/src/types/context.rs
+++ b/crates/red_knot_python_semantic/src/types/context.rs
@@ -126,8 +126,8 @@ impl<'db> InferContext<'db> {
         });
     }
 
-    pub(super) fn set_in_no_type_check(&mut self, no_type_check: InNoTypeCheck) {
-        self.no_type_check = no_type_check;
+    pub(super) fn record_no_type_check_block_entered(&mut self) {
+        self.no_type_check = InNoTypeCheck::Yes;
     }
 
     fn is_in_no_type_check(&self) -> bool {
diff --git a/crates/red_knot_python_semantic/src/types/infer.rs b/crates/red_knot_python_semantic/src/types/infer.rs
index fc4970462..27132d97c 100644
--- a/crates/red_knot_python_semantic/src/types/infer.rs
+++ b/crates/red_knot_python_semantic/src/types/infer.rs
@@ -72,7 +72,7 @@ use crate::unpack::Unpack;
 use crate::util::subscript::{PyIndex, PySlice};
 use crate::Db;
 
-use super::context::{InNoTypeCheck, InferContext, WithDiagnostics};
+use super::context::{InferContext, WithDiagnostics};
 use super::diagnostic::{
     report_index_out_of_bounds, report_invalid_exception_caught, report_invalid_exception_cause,
     report_invalid_exception_raised, report_non_subscriptable,
@@ -1033,7 +1033,7 @@ impl<'db> TypeInferenceBuilder<'db> {
             // and, if so, suppress any errors.
             if let Type::FunctionLiteral(function) = ty {
                 if function.is_known(self.db(), KnownFunction::NoTypeCheck) {
-                    self.context.set_in_no_type_check(InNoTypeCheck::Yes);
+                    self.context.record_no_type_check_block_entered();
                 }
             }
         }
```

But the advantage of your current version is that it's more extensible, so I don't have a strong opinion